staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/ratcheting.go

```diff
@@ -0,0 +1,336 @@
+			if !ok {
+				return false
+			} else if !childInvocation.IsNoopLazy() {
+				return false
+			}
+		}
+
+		return true
+	}
+
+	oldAsMap, oldIsMap := r.thunk.oldValue.(map[string]interface{})
+	newAsMap, newIsMap := r.value.(map[string]interface{})
+
+	if oldIsMap != newIsMap {
+		return false
+	} else if oldIsMap {
+		if len(oldAsMap) != len(newAsMap) {
+			return false
+		}
+
+		for k := range oldAsMap {
+			// Query for child
+			childInvocation, ok := r.children[k]
+			if !ok {
+				// Either key in old not in new, or validator short
+				// circuited
+				_, newAlsoHasKey := newAsMap[k]
+				if newAlsoHasKey {
+					// Child invocation is missing due to short circuit
+					return reflect.DeepEqual(r.thunk.oldValue, r.value)
+				}
+
+				// Child invocation is missing due to key change
+				return false
+			} else if !childInvocation.IsNoopLazy() {
+				return false
+			}
+		}
+
+		return true
+	}
+
+	return reflect.DeepEqual(r.thunk.oldValue, r.value)
+}
+
+func newRatchetingValueValidator(oldValue interface{}, args schemaArgs, parent *ratchetingValueValidator, keyInParent interface{}) *ratchetingValueValidator {
+	return &ratchetingValueValidator{
+		oldValue:    oldValue,
+		schemaArgs:  args,
+		parent:      parent,
+		keyInParent: keyInParent,
+	}
+}
+
+func (r *ratchetingValueValidator) validateOption(value interface{}) validate.Option {
+	return func(svo *validate.SchemaValidatorOptions) {
+		svo.NewValidatorForField = func(field string, schema *spec.Schema, rootSchema interface{}, root string, formats strfmt.Registry, opts ...validate.Option) validate.ValueValidator {
+			return r.SubPropertyValidator(value, field, schema, rootSchema, root, formats, opts...)
+		}
+		svo.NewValidatorForIndex = func(index int, schema *spec.Schema, rootSchema interface{}, root string, formats strfmt.Registry, opts ...validate.Option) validate.ValueValidator {
+			return r.SubIndexValidator(value, index, schema, rootSchema, root, formats, opts...)
+		}
+	}
+}
+
+func (r *ratchetingValueValidator) Applies(value interface{}, kind reflect.Kind) bool {
+	return true
+}
+
+func (s *ratchetingValueValidator) SetPath(path string) {
+	s.schemaArgs.Path = path
+}
+
+// Validate validates the value.
+func (r *ratchetingValueValidator) Validate(value interface{}) *validate.Result {
+	r.currentInvocation = &ratchetInvocationOptions{
+		value:            value,
+		thunk:            r,
+		children:         map[interface{}]*ratchetInvocationOptions{},
+		comparisonResult: nil,
+	}
+	defer func() {
+		r.currentInvocation = nil
+	}()
+
+	if r.parent != nil {
+		r.parent.currentInvocation.children[r.keyInParent] = r.currentInvocation
+	}
+
+	opts := append([]validate.Option{
+		r.validateOption(value),
+	}, r.Options...)
+
+	s := validate.NewSchemaValidator(r.Schema, r.Root, r.Path, r.KnownFormats, opts...)
+
+	res := s.Validate(value)
+
+	if res.IsValid() {
+		return res
+	}
+
+	// Current policy is to ratchet errors if the comparison is a noop
```

Comment:
what does "current policy" mean? what is a noop?

Reply:
rephrased this comment